### ISSUE TITLE
Should `ioHintSet` support additional hints as type methods?

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#20346]
The `ioHintSet` type currently supports the following hints:
- sequential
- random
- prefetch
- mmap

However; there are a few other hints supported by the runtime which were not included in the original proposal (here: https://github.com/chapel-lang/chapel/issues/20141):
- QIO_HINT_OWNED
- QIO_METHOD_PREADPWRITE
- QIO_HINT_PARALLEL
- QIO_CH_ALWAYS_UNBUFFERED

These are currently being imported and used via the `ioHintSet.direct()` method. For example:
https://github.com/chapel-lang/chapel/blob/a104ef5150cd87eadd3ee0c5be9b0f1a17076314/modules/standard/Subprocess.chpl#L584-L587
___

Should `ioHintSet` have type methods for these hints instead of requiring them to be used via the `direct` method?

Specifically, should we add any of the following to the `ioHintSet` type:

- ioHintSet.owned
- ioHintSet.pReadPWrite
- ioHintSet.parallel
- ioHintSet.unbuffered

**Comment by lydia-duncan** [comment]
Thanks Jeremiah!  Here's some details that will probably be notable to other people:
- test/io/tzakian/recordReader/test.chpl's use of `QIO_METHOD_PREADPWRITE` is explicitly called out as for when `mmap` is not used.  I seem to recall @jhh67 proposing a `noMmap` enum constant option, and it seems like that would be a good replacement for this one.
- The shootout benchmarks (which have typically been near and dear to Brad's heart) use `QIO_HINT_PARALLEL` and `QIO_CH_ALWAYS_UNBUFFERED`.  We should have some equivalent so that they can use a user-facing feature without losing performance
- The fileIO primer (which will probably be rewritten when encoders/decoders come online) used `QIO_HINT_PARALLEL` and I encouraged Jeremiah to drop it from there since it wasn't user-facing.  If we make a user-facing version, we can re-add it there (but I don't know that we have to for pedagogical reasons)

**Comment by bradcray** [comment]
> The shootout benchmarks (which have typically been near and dear to Brad's heart) use QIO_HINT_PARALLEL and QIO_CH_ALWAYS_UNBUFFERED. We should have some equivalent so that they can use a user-facing feature without losing performance

Totally agree with this.  Just to double down on it:  Some members of the IO team have heard me say this at one point or another, but I think the use of IO in the shootout benchmarks is particularly ugly / non-Chapeltastic, and I think it would be great to create versions of them that look beautiful and get great performance (both because they're high-profile codes and because it would suggest to me that users could use reasonable features to get competitive performance).  We'll also need to be careful in updating these tests as we'll need to update the versions on the live site as these features are deprecated and removed, again, ideally without losing any performance and only improving clarity.

I know @jhh67 changed some IO implementation details recently that made me wonder whether some of the hints/tricks we use in the shootouts might be less important, but I haven't had the chance to try that out yet (and there's also always the question of whether a change—or lack thereof—on our system will or will not also show up as a change on the system used for the benchmarks...  we used to have a pretty good clone of the system that we could try things out on, but haven't for awhile.  And even then, our IO performance often seemed hard to reconcile with what the live site was seeing.  @ronawho probably remembers details there better than I do at this point).

**Comment by jeremiah-corrado** [comment]
`QIO_HINT_OWNED` method name:

- 🎉 `owned`
- 😃 `callerOwned`
- 🚀 `persistant`
- 👀 - something else

**Comment by jeremiah-corrado** [comment]
`QIO_HINT_PARALLEL` method name:

- 🎉 `parallel`
- 😃 `threadSafe`
- 🚀 `multiThreaded`
- 👀 - something else

**Comment by jeremiah-corrado** [comment]
`QIO_CH_ALWAYS_UNBUFFERED` method name:

- 🎉 `unbuffered`
- 😃 `alwaysUnbuffered`
- 👀 - something else

**Comment by jeremiah-corrado** [comment]
Looking into `QIO_METHOD_PREADPWRITE` a bit closer, I am somewhat hesitant to include a method for it in `ioHintSet`. As far as I can tell, this is more of an internal implementation detail in the runtime. Pinging @jhh67 to see if I am misunderstanding.

It is only exposed in one test currently: `io/tzakian/recordReader/test.chpl`, so I wouldn't have a huge problem leaving that one as: `ioHintSet.direct(QIO_METHOD_PREADPWRITE)`.
___

what to do with `QIO_METHOD_PREADPWRITE `?

- 😕 - don't include it as a type method
- 🎉 - include as `preadpwrite`
- 👀 - include with different name

**Comment by benharsh** [comment]
I want to put an idea out there for QIO_METHOD_PREADPWRITE: Could we have two methods along the lines of "pread" and "pwrite"? They would both map to the same value, but it might be a bit clearer when reading code what the user actually wants when dealing with a reader/writer.

**Comment by mppf** [comment]
I don't think the set of hints discussed here is the set I would add. I know you are trying to ask about them individually, but there are others that aren't here.

See https://github.com/chapel-lang/chapel/blob/cd06ad7c774b995c01c8908f21f0d85627092e78/runtime/include/qio/qio.h#L246-L279

Note that `owned` has the opposite meaning from `callerOwned` -- it's more like "QIO should close it when the file closes".

My initial reaction to this question overall is that none of these should be supported as type methods. However, the situation with the shootouts seems a bit untenable. In particular, I would like to delete `QIO_CH_ALWAYS_UNBUFFERED` from the implementation if I can... I don't think that makes much sense in the long term.

I'd have to look at the implementation to know for sure but IIRC `QIO_HINT_PARALLEL` just activates mmap. If that's the case, the existing mmap hint should be good enough.

**Comment by mppf** [comment]
(I forgot to say in my previous response, but a hint that activates `O_DIRECT` might be useful, and `QIO_HINT_DIRECT` does that now, but we have a terminology conflict with `.direct(int)` used above)

**Comment by jeremiah-corrado** [comment]
> I want to put an idea out there for QIO_METHOD_PREADPWRITE: Could we have two methods along the lines of "pread" and "pwrite"? They would both map to the same value, but it might be a bit clearer when reading code what the user actually wants when dealing with a reader/writer.

That's a good idea. My only concern would be that the user might expect the following to either produce an error or not invoke the `pwrite` hint:
```chapel
var f = open("path/to/file.txt", iomode.cw, hints = ioHintSet.pread);
var w = f.writer()
...
```

> I don't think the set of hints discussed here is the set I would add. I know you are trying to ask about them individually, but there are others that aren't here.

That's a good point. This issue was specifically focused on the cases where `ioHintSet.direct()` is being invoked in modules or tests. I was thinking in terms of 2.0 atm, i.e., getting rid of those calls wherever possible, then potentially adding other hints later on.

> (I forgot to say in my previous response, but a hint that activates O_DIRECT might be useful, and QIO_HINT_DIRECT does that now, but we have a terminology conflict with .direct(int) used above)

Thanks for pointing that out! I was unaware of that hint when I created that method. Hopefully we can remove all the calls to `direct` and then remove that method.

**Comment by bradcray** [comment]
> 🎉 owned

Note that `owned` is a keyword in Chapel, so can't be an identifier name (well, without playing special games in the parser, which I don't think we'd want to do here).

**Comment by jeremiah-corrado** [comment]
> Note that owned is a keyword in Chapel, so can't be an identifier name (well, without playing special games in the parser, which I don't think we'd want to do here).

> Note that owned has the opposite meaning from callerOwned -- it's more like "QIO should close it when the file closes".

Would something like `fileOwned` be better?

**Comment by jhh67** [comment]
I'd like to make the following suggestions. 

QIO_HINT_OWNED -- This isn't a hint, it's a flag/directive that the associated file descriptor or FILE* should be closed when the file is closed. I think it should only be passed to `openfd` and `openfp` and could simply be a boolean argument to them.

QIO_METHOD_PREADPWRITE -- As Ben H. points out this is probably better implemented as `pread` and `pwrite` methods on the channel.

QIO_HINT_PARALLEL -- This controls whether or not mmap is used, should be replaced by hints to use mmap or not.

QIO_CH_ALWAYS_UNBUFFERED -- Michael says this isn't needed, but it seems useful to me to specify that I/O should be unbuffered. It eliminates buffer space and copies to/from the buffer. 

QIO_HINT_DIRECT -- This also seems useful.

It would be good to run the relevant shootout benchmarks with different combinations of hints (including none) to see what effect they have.

**Comment by mppf** [comment]
Just pointing out that `QIO_HINT_DIRECT` and `QIO_CH_ALWAYS_UNBUFFERED` both remove buffering. (I'm not so sure `QIO_CH_ALWAYS_UNBUFFERED` does what we think it does, but for now let's assume it does). `QIO_HINT_DIRECT` activates `O_DIRECT` which bypasses the OS "page cache" which buffers file data in memory. `QIO_CH_ALWAYS_UNBUFFERED` should disable the channel buffer.


Part of the reason I would prefer to remove `QIO_CH_ALWAYS_UNBUFFERED` is that my view is that buffering is pretty much the entire point of a `qio_channel_t`/`channel`/`fileReader`/`fileWriter`. So if it is not doing buffering, is it doing anything useful to describe the I/O in that way?

In C you get buffering when doing I/O with a `FILE*` (e.g. `fwrite`) but not with a file descriptor (e.g. `write` system call). Which leads me to wonder if we should have a totally different low-level write mechanism outside of the `fileReader`/`fileWriter`. Or, if it is better to have it be a configurable property of a `fileReader`/`fileWriter` similar to `QIO_CH_ALWAYS_UNBUFFERED` but with a different interface.

One concern I have about treating "should it buffer at all" as a hint is that some of the methods like `mark` `revert` `commit` will not work if buffering is disabled. It seems to me that moves it beyond a "hint" and to some sort of configurable functionality. (I.e. so maybe it would be better as a `buffering: bool` argument to the functions that return a new channel like `file.reader`).

**Comment by jhh67** [comment]
I like the idea of an "unbuffered channel" being a different type for which some methods are not supported, e.g. `mark`.

**Comment by jeremiah-corrado** [comment]
> I'd like to point out that QIO_HINT_DIRECT and QIO_CH_ALWAYS_UNBUFFERED both remove buffering. 

> I like the idea of an "unbuffered channel" being a different type for which some methods are not supported, e.g. mark.

If non-buffering IO was done with a separate type/interface, would it still be useful to include`QIO_HINT_DIRECT` in `ioHintSet`?

**Comment by mppf** [comment]
> If non-buffering IO was done with a separate type/interface, would it still be useful to include `QIO_HINT_DIRECT` in `ioHintSet`?

Maybe? It controls the buffering at a different layer (the OS rather than the application). Note though that `O_DIRECT` is not particularly easy to use in linux and we might be better off *not* trying to expose it through the channel interface. Perhaps, if we had a separate un-buffered I/O interface, it could support `O_DIRECT` as well as skipping the buffering that the qio/channel code normally does.

I guess to me it depends on whether or not it will be common that users want access to `O_DIRECT`. It could be a feature request to make it more convenient, say, if we don't support it from the outset. For our Chapel 2.0 effort, I don't think we should do anything at all with `O_DIRECT`, other than consider renaming `ioHintSet.direct()` for future-proofing and to avoid confusion.

**Comment by jeremiah-corrado** [comment]
Ok that makes sense, thanks! 

Hopefully we can get rid of all the uses of `direct` and remove it entirely. If not, I can rename it to something like `fromInt`.

**Comment by jeremiah-corrado** [comment]
So overall, it sounds like none of the hints/methods brought up in this issue should be included in `ioHintSet`:

 - `QIO_HINT_OWNED` : would be better as a boolean argument to `openfd` and `openfp`
 - `QIO_HINT_PARALLEL `: is already handled by the `mmap` hint
 - `QIO_CH_ALWAYS_UNBUFFERED`: unbuffered io should potentially have its own io types/interface, or it should be a boolean argument on functions that return a file
 - `QIO_METHOD_PREADPWRITE `: should be implemented as two IO methods: `pread` and `pwrite`

I plan to open separate issues for each of the above (except `PARALLEL`) unless anyone has objections and thinks that one or more of these should be included in `ioHintSet`.

**Comment by mppf** [comment]
> `QIO_METHOD_PREADPWRITE `: should be implemented as two IO methods: `pread` and `pwrite`

I thought the proposal was to have `ioHintSet.pread` and `ioHintSet.pwrite`. Another way to handle this though is to add a hint along the lines of `notMmap`.

**Comment by jeremiah-corrado** [comment]
Oh, maybe I'm misunderstanding, I was looking at:

> QIO_METHOD_PREADPWRITE -- As Ben H. points out this is probably better implemented as pread and pwrite methods on the channel.

So I figured I had initially misinterpreted Ben's message:

> I want to put an idea out there for QIO_METHOD_PREADPWRITE: Could we have two methods along the lines of "pread" and "pwrite"? They would both map to the same value, but it might be a bit clearer when reading code what the user actually wants when dealing with a reader/writer.

@benharsh were you advocating for new iohints or new methods?

**Comment by benharsh** [comment]
@jeremiah-corrado : I'm sorry, I was advocating for *type methods* on ioHintSet. I latched on to this terminology from the Issue's main text:

> Should ioHintSet have type methods for these hints instead of requiring them to be used via the direct method?
> 
> Specifically, should we add any of the following to the ioHintSet type:

I am not proposing that we add these as methods to ``channel``.

**Comment by jeremiah-corrado** [comment]
No worries, thanks for the clarification!

> Another way to handle this though is to add a hint along the lines of `notMmap`.

I also like this idea. Other possible names: `offset`, `noMmap`

**Comment by mppf** [comment]
> I also like this idea. Other possible names: `offset`, `noMmap`

I'm not sure `offset` would work for me as a name.

`readWrite` or `readAndWrite` or something like that could also be the names, potentially.

**Comment by jeremiah-corrado** [comment]
Ok, maybe a poll for the name of the `QIO_METHOD_PREADPWRITE` hint(s) would be a good idea:

- 😄 - `pread` and `pwrite` (both pointing to `QIO_METHOD_PREADPWRITE`)
- 🎉 - `noMmap`
- ❤️ - `readWrite`
- 🚀 - `readAndWrite`
- 👀 - something else...

**Comment by ronawho** [comment]
> @jhh67 changed some IO implementation details recently that made me wonder whether some of the hints/tricks we use in the shootouts might be less important, but I haven't had the chance to try that out yet (and there's also always the question of whether a change—or lack thereof—on our system will or will not also show up as a change on the system used for the benchmarks... we used to have a pretty good clone of the system that we could try things out on, but haven't for awhile. And even then, our IO performance often seemed hard to reconcile with what the live site was seeing. @ronawho probably remembers details there better than I do at this point).

I know the conversation has moved on from this point, but I'm just getting around to responding post vacation. Feel free to hide this comment if it gets in the way

I think you're referring specifically to https://github.com/chapel-lang/chapel/issues/9077 where after a problem size increase the old shootout system didn't have enough RAM to keep both the program data and output data in memory at the same time, so we'd end up swapping and hurting performance. This was after a problem size increase and was seen in our testing infra because we save the output in order to do verification but the actual shootout site will discard output for performance runs. I don't really think this is an issue anymore since the shootout machine was upgraded from a system with 4G of ram to a system with 16G and no problem sizes have increased.

We have a somewhat decent approximation of this machine in mnlchapvm05. It has the same core count and memory, but it's a VM so you never know how reliable performance figures are, especially for I/O. Generally speaking though, I saw rather stable performance and performance that is close to the real shootout machine, so I think it'd be a decent place to test I/O changes that impact the shootout benchmarks.

**Comment by jeremiah-corrado** [comment]
- `QIO_METHOD_PREADPWRITE` was resolved by: #20468.
- `QIO_HINT_OWNED` moved to #20430 
- `QIO_CH_ALWAYS_UNBUFFERED` moved to #20432 
- `QIO_HINT_PARALLEL` is already handled by `ioHintSet.mmap`
